src/main/java/io/snyk/agent/logic/ReportingWorker.java

```diff
@@ -132,6 +134,34 @@ void sendIfNecessary(Supplier<UseCounter.Drain> drainMaker) throws IOException {
         this.lastSuccessfulBeacon = now;
     }
 
+    private void setShutdownHook() {
+        try {
+            log.info("attempting to set up a shutdown hook that drains events one last time");
+            final Thread shutdownHook = new Thread(() -> safeSendIfNecessary(LandingZone.SEEN_SET::drain));
+            Runtime.getRuntime().addShutdownHook(shutdownHook);
+            log.info("shutdown hook set up successfully");
+        } catch (IllegalArgumentException e) {
+            // If the specified hook has already been registered,
+            // or if it can be determined that the hook is already running or has already been run.
+            // so either way we're set?
+        } catch (IllegalStateException e) {
+            // If the virtual machine is already in the process of shutting down
+            // this means we're okay with not setting up that hook
+        } catch (SecurityException e) {
+            // If a security manager is present and it denies RuntimePermission("shutdownHooks")
+            // not much to do about this one except log?
+            log.warn("failed to set a shutdownHook due to denied permission");
+        }
+    }
+
+    private void safeSendIfNecessary(Supplier<UseCounter.Drain> drainMaker) {
+        try {
+            sendIfNecessary(LandingZone.SEEN_SET::drain);
```

Comment:
The decision as to whether it is necessary to send anything is made inside `sendIfNecessary`. It's very likely that this call will decide that the timer hasn't yet expired, so to do nothing, and we'll lose the events.

Reply:
good catch, thanks!
I guess I can bypass that with one of the following:
1. add some boolean (`isShutdown`) which will bypass the timer - this is a bit ugly in my opinion
2. temper with the timer in `safeSendIfNecessary` prior to calling `sendIfNecessary` - sounds pretty bad since we're overriding that internal state for different meanings/needs
3. extract the "draining" logic to a different function and invoke it specifically in the `shutdownHook` context. this seems to make the most sense to me.

what do you think?

Reply:
I'd go for another function. There's no real reason for it to be inline. We could even do the `drain` outside the function, to make that bit clearer.

Reply:
pushed a naive implementation for the separation, not sure I got the "`drain` outside" part right though. see if it makes sense please